traversal/producer/GraphProducer.java

```diff
@@ -46,91 +47,102 @@
-            }
+        isInitialised = true;
+        if (runningJobs.isEmpty()) done(queue);
+    }
+
+    private synchronized void distribute(Queue<VertexMap> queue, int request) {
+        if (isDone.get()) return;
+        int requestSplitMax = (int) Math.ceil((double) request / runningJobs.size());
+        int requestSent = 0;
+        for (ResourceIterator<VertexMap> iterator : runningJobs.keySet()) {
+            int requestSplit = Math.min(requestSplitMax, request - requestSent);
+            runningJobs.computeIfPresent(iterator, (iter, asyncJob) -> asyncJob.thenRunAsync(
+                    () -> job(queue, iter, requestSplit), forkJoinPool()
+            ));
+            requestSent += requestSplit;
+            if (requestSent == request) break;
         }
     }
 
-    private synchronized int take(ResourceIterator<VertexMap> iterator) {
-        int count = iteratorRequested.get(iterator);
-        iteratorRequested.put(iterator, 0);
-        return count;
+    private synchronized void transition(Queue<VertexMap> queue, ResourceIterator<VertexMap> iterator, int unfulfilled) {
+        if (iterator.hasNext()) return;
+        if (runningJobs.containsKey(iterator) && start.hasNext()) replace(queue, iterator, unfulfilled);
+        else if (!runningJobs.isEmpty() && unfulfilled > 0) distribute(queue, unfulfilled);
+        else if (runningJobs.isEmpty()) done(queue);
     }
 
-    private synchronized void compensate(Queue<VertexMap> queue, ResourceIterator<VertexMap> iterator, int requested) {
-        iteratorJobs.remove(iterator);
-        int toCompensate = requested + take(iterator);
-        if (!iterator.hasNext()) iteratorRequested.remove(iterator);
-        if (toCompensate > 0) produce(queue, toCompensate);
+    private synchronized void replace(Queue<VertexMap> queue, ResourceIterator<VertexMap> iterator, int unfulfilled) {
+        runningJobs.remove(iterator);
+        ResourceIterator<VertexMap> newIter =
+                new GraphIterator(graphMgr, start.next(), procedure, params, filter).distinct(produced);
+        CompletableFuture<Void> asyncJob = unfulfilled > 0
+                ? CompletableFuture.runAsync(() -> job(queue, newIter, unfulfilled), forkJoinPool())
+                : CompletableFuture.completedFuture(null);
+        runningJobs.put(newIter, asyncJob);
     }
 
-    private void produceAsync(Queue<VertexMap> queue, ResourceIterator<VertexMap> iterator) {
+    private void job(Queue<VertexMap> queue, ResourceIterator<VertexMap> iterator, int request) {
         try {
-            int requested;
-            int produced;
-            do {
-                requested = take(iterator);
-                produced = 0;
-                for (; produced < requested && iterator.hasNext(); produced++)
+            int i = 0;
+            if (runningJobs.containsKey(iterator)) {
+                for (; i < request && iterator.hasNext() && !isDone.get(); i++) {
                     queue.put(iterator.next());
```

Comment:
Is it possible that `done()` is called after `isDone.get()` check?

Reply:
it is possible, but only when there has been an error. I don't think it's possible if the producer was ended correctly?